Blog for KEP-5836: Scheduler Preemption for In-Place Pod Resize (alpha) (#56461)
kubernetes/website: content/en/blog/_posts/2026-xx-xx-scheduler-preemption-for-in-place-pod-resize.md

---
layout: blog
title: "Kubernetes v1.37: Scheduler Preemption for In-Place Pod Resize (Alpha)"
slug: scheduler-preemption-in-place-pod-resize-alpha
draft: true

> date: XXXX-XX-XX

Reviewer: Nit: Not needed, the date field will be added once the draft gets merged in a separate publication PR

```suggestion

```

Author: done!

Author: actually, removing the date field resulted int his error in CI:

2:24:16 PM: ERROR the "date" front matter field is not a parsable date: see /opt/build/repo/content/en/blog/_posts/2026-xx-xx-scheduler-preemption-for-in-place-pod-resize.md

## The "Deferred" Resize Challenge

To understand why this preemption mechanism is needed, it is helpful to look at how Kubernetes handles running pod resizing. When a user or controller (such as the Vertical Pod Autoscaler) updates the resource requests of an active container, the Kubelet evaluates whether the underlying node has enough spare allocatable capacity to fulfill the increase. 

If the node's resources are fully utilized and cannot satisfy the new limits, the Kubelet sets the container's `resizeStatus` (reported in the Pod's `status.containerStatuses[]`) to `Deferred`. Unlike an `Infeasible` resize request (which is immediately rejected because it exceeds physical machine boundaries, namespace limit ranges, or admission quotas) a `Deferred` status indicates that the request is valid but is temporarily unable to be actuated, waiting until node capacity becomes available.

> Before the introduction of this preemption mechanism, a pod's in-place resize scale-up request could become permanently blocked if the node was heavily utilized. If a critical application (such as an in-memory database or a real-time web server) required more memory to prevent an imminent out-of-memory (OOM) crash, and the node lacked free capacity, the resize remained `Deferred`. 

Reviewer 2: ```suggestion
Before the introduction of this preemption mechanism, a pod's in-place resize scale-up request could become permanently blocked if the node was heavily utilized. Even though a critical application (such as an in-memory database or a real-time web server) required more memory to prevent an imminent out-of-memory (OOM) crash, and the node lacked free capacity, the resize remained `Deferred`. 
```
Or "Even when"

Author: changed to "even when". Thanks!

* **Maximum Size**: The array can contain a maximum of **20 items**.
* **Value Format**: Each string in the list must conform strictly to the standard **Kubernetes label key format** (up to 63 characters with optional DNS subdomain prefixes).

## Prerequisites and Limitations

To utilize Scheduler Preemption for In-Place Pod Resize:
* Both the base `InPlacePodVerticalScaling` and the `InPlacePodVerticalScalingSchedulerPreemption` feature gates must be enabled across all control plane components (`kube-apiserver`, `kube-scheduler`) and the `kubelet`.
* Preemption is strictly scoped to the same node where the deferred pod is running. If a node cannot accommodate the resize even after evicting all eligible lower-priority workloads, the resize remains in the `Deferred` state.

## Getting Involved

This feature represents a major step forward for resource scheduling, bringing enterprise-grade density control and workload prioritization to dynamic resource scaling. We invite cluster operators, platform architects, and developers to enable the `InPlacePodVerticalScalingSchedulerPreemption` feature gate in their testing environments and share feedback.

> If you want to share your experience with this feature, please get in touch with the community via [SIG Scheduling](https://github.com/kubernetes/community/tree/master/sig-scheduling) or [SIG Node](https://github.com/kubernetes/community/tree/master/sig-node) channels!

Reviewer: Optionally: use https://www.kubernetes.dev/community/community-groups/sigs/scheduling and 
https://www.kubernetes.dev/community/community-groups/sigs/node as the URL for SIG, more user friendly
```suggestion
If you want to share your experience with this feature, please get in touch with the community via [SIG Scheduling](https://www.kubernetes.dev/community/community-groups/sigs/scheduling) or [SIG Node](https://www.kubernetes.dev/community/community-groups/sigs/node/).
```

Author: done
